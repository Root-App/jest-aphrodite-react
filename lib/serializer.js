'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
var cssUtil_1 = require('./cssUtil');
var reactUtils_1 = require('./reactUtils');
var replaceClassNames_1 = require('./replaceClassNames');
function createSerializer(getStyleSheetTestUtils, _a) {
  var _b = _a === void 0 ? {} : _a,
    _c = _b.removeVendorPrefixes,
    removeVendorPrefixes = _c === void 0 ? false : _c,
    classNameReplacer = _b.classNameReplacer;
  function test(val) {
    return (
      val &&
      !val.withStyles &&
      (val.$$typeof === Symbol.for('react.test.json') ||
        (val instanceof HTMLElement && !isBeingSerialized(val)))
    );
  }
  function print(val, printer) {
    var nodes = reactUtils_1.getNodes(val);
    nodes.forEach(function(node) {
      return (node.withStyles = true);
    });
    var selectors = cssUtil_1.getSelectors(nodes);
    var styles = cssUtil_1.getStyles(
      selectors,
      getStyleSheetTestUtils,
      removeVendorPrefixes,
    );
    var printedVal = printer(val);
    if (styles) {
      return replaceClassNames_1.replaceClassNames(
        selectors,
        styles,
        printedVal,
        classNameReplacer,
      );
    } else {
      return printedVal;
    }
  }
  return {
    test: test,
    print: print,
  };
}
exports.createSerializer = createSerializer;
function isBeingSerialized(node) {
  var currentNode = node;
  while (currentNode) {
    if (currentNode.withStyles) {
      return true;
    }
    currentNode = currentNode.parentNode;
  }
  return false;
}
// doing this to make it easier for users to mock things
// like switching between development mode and whatnot.
var getAphroditeStyleSheetTestUtils = function(useImportant) {
  return function() {
    return require('aphrodite' + (useImportant ? '' : '/no-important'))
      .StyleSheetTestUtils;
  };
};
exports.serializer = function(useImportant) {
  return createSerializer(getAphroditeStyleSheetTestUtils(useImportant));
};
