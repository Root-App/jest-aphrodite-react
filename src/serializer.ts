import { StyleSheetTestUtils } from 'aphrodite';
import { ReactTestRendererJSON } from 'react-test-renderer';
import { getSelectors, getStyles } from './cssUtil';
import { getNodes } from './reactUtils';
import { ClassNameReplacer, replaceClassNames } from './replaceClassNames';

export interface SerializerOptions {
  removeVendorPrefixes?: boolean;
  classNameReplacer?: ClassNameReplacer;
}

interface TrackedHTMLElement extends HTMLElement {
  withStyles?: boolean;
}

interface TrackedReactTestRendererJSON extends ReactTestRendererJSON {
  withStyles?: boolean;
}

type TrackedValue = TrackedHTMLElement | TrackedReactTestRendererJSON;

export function createSerializer(
  getStyleSheetTestUtils: () => typeof StyleSheetTestUtils,
  { removeVendorPrefixes = false, classNameReplacer }: SerializerOptions = {},
): jest.SnapshotSerializerPlugin {
  function test(val: any) {
    return (
      val &&
      !val.withStyles &&
      (val.$$typeof === Symbol.for('react.test.json') ||
        (val instanceof HTMLElement && !isBeingSerialized(val)))
    );
  }

  function print(val: TrackedValue, printer: (val: any) => string) {
    const nodes = getNodes(val);
    nodes.forEach((node: TrackedValue) => (node.withStyles = true));

    const selectors = getSelectors(nodes);
    const styles = getStyles(
      selectors,
      getStyleSheetTestUtils,
      removeVendorPrefixes,
    );

    const printedVal = printer(val);
    delete val.withStyles;
    nodes.forEach((node: TrackedValue) => delete node.withStyles);

    if (styles) {
      return replaceClassNames(
        selectors,
        styles,
        printedVal,
        classNameReplacer,
      );
    } else {
      return printedVal;
    }
  }

  return {
    test,
    print,
  };
}

function isBeingSerialized(node: TrackedHTMLElement) {
  let currentNode = node;

  while (currentNode) {
    if (currentNode.withStyles) {
      return true;
    }
    currentNode = currentNode.parentNode as TrackedHTMLElement;
  }
  return false;
}

// doing this to make it easier for users to mock things
// like switching between development mode and whatnot.
const getAphroditeStyleSheetTestUtils = (
  useImportant: boolean,
) => (): typeof StyleSheetTestUtils =>
  require(`aphrodite${useImportant ? '' : '/no-important'}`)
    .StyleSheetTestUtils;

export const serializer = (useImportant: boolean) =>
  createSerializer(getAphroditeStyleSheetTestUtils(useImportant));
