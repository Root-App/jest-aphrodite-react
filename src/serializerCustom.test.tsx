import { css, StyleSheet, StyleSheetTestUtils } from 'aphrodite';
import * as enzyme from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import * as reactTestRenderer from 'react-test-renderer';
import { createSerializer } from './serializer';

expect.addSnapshotSerializer(
  createSerializer(() => StyleSheetTestUtils, { removeVendorPrefixes: true }),
);

const Wrapper: React.SFC = props => {
  const styles = StyleSheet.create({
    wrapper: {
      display: 'flex',
      transform: 'translate(0, 0)',
    },
  });
  return <section className={css(styles.wrapper)} {...props} />;
};

const Title: React.SFC = props => {
  const styles = StyleSheet.create({
    title: {
      color: 'palevioletred',
      fontSize: '1.5em',
      textAlign: 'center',
    },
  });
  return <h1 className={css(styles.title)} {...props} />;
};

test('removes vendor prefixed rules from react-test-renderer', () => {
  const tree = reactTestRenderer
    .create(
      <Wrapper>
        <Title>
          Hello World, this is my first component styled with aphrodite!
        </Title>
      </Wrapper>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('removes vendor prefixed rules from enzyme', () => {
  const ui = (
    <Wrapper>
      <Title>
        Hello World, this is my first component styled with aphrodite!
      </Title>
    </Wrapper>
  );

  const enzymeMethods = ['shallow', 'mount', 'render'];
  enzymeMethods.forEach(method => {
    const tree = (enzyme as any)[method](ui);
    expect(toJson(tree)).toMatchSnapshot(`enzyme.${method}`);
  });
});

test('removes vendor prefixes inside mediaQueries', () => {
  const styles = StyleSheet.create({
    root: {
      '@media(min-width: 200px)': {
        transform: 'translate(0, 0)',
      },
      '@supports(display: grid)': {
        display: 'flex',
      },
    },
  });
  const tree = reactTestRenderer
    .create(<div className={css(styles.root)} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
