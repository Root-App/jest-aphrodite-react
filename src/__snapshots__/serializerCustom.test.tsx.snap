// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removes vendor prefixed rules: enzyme.mount 1`] = `
.wrapper {
  display: flex !important;
  transform: translate(0, 0) !important;
}

.title {
  color: palevioletred !important;
  font-size: 1.5em !important;
  text-align: center !important;
}

<Wrapper>
  <section
    className="wrapper"
  >
    <Title>
      <h1
        className="title"
      >
        Hello World, this is my first component styled with aphrodite!
      </h1>
    </Title>
  </section>
</Wrapper>
`;

exports[`removes vendor prefixed rules: enzyme.render 1`] = `
.wrapper {
  display: flex !important;
  transform: translate(0, 0) !important;
}

.title {
  color: palevioletred !important;
  font-size: 1.5em !important;
  text-align: center !important;
}

<section
  class="wrapper"
>
  <h1
    class="title"
  >
    Hello World, this is my first component styled with aphrodite!
  </h1>
</section>
`;

exports[`removes vendor prefixed rules: enzyme.shallow 1`] = `
.wrapper {
  display: flex !important;
  transform: translate(0, 0) !important;
}

<section
  className="wrapper"
>
  <Title>
    Hello World, this is my first component styled with aphrodite!
  </Title>
</section>
`;

exports[`removes vendor prefixed rules: react-test-renderer 1`] = `
.wrapper {
  display: flex !important;
  transform: translate(0, 0) !important;
}

.title {
  color: palevioletred !important;
  font-size: 1.5em !important;
  text-align: center !important;
}

<section
  className="wrapper"
>
  <h1
    className="title"
  >
    Hello World, this is my first component styled with aphrodite!
  </h1>
</section>
`;

exports[`removes vendor prefixed rules: react-testing-library 1`] = `
.wrapper {
  display: flex !important;
  transform: translate(0, 0) !important;
}

.title {
  color: palevioletred !important;
  font-size: 1.5em !important;
  text-align: center !important;
}

<section
  class="wrapper"
>
  <h1
    class="title"
  >
    Hello World, this is my first component styled with aphrodite!
  </h1>
</section>
`;

exports[`removes vendor prefixes inside mediaQueries: enzyme.mount 1`] = `
@media (min-width: 200px) {
  .root {
    transform: translate(0, 0) !important;
  }
}

@supports (display: grid) {
  .root {
    display: flex !important;
  }
}

<div
  className="root"
/>
`;

exports[`removes vendor prefixes inside mediaQueries: enzyme.render 1`] = `
@media (min-width: 200px) {
  .root {
    transform: translate(0, 0) !important;
  }
}

@supports (display: grid) {
  .root {
    display: flex !important;
  }
}

<div
  class="root"
/>
`;

exports[`removes vendor prefixes inside mediaQueries: enzyme.shallow 1`] = `
@media (min-width: 200px) {
  .root {
    transform: translate(0, 0) !important;
  }
}

@supports (display: grid) {
  .root {
    display: flex !important;
  }
}

<div
  className="root"
/>
`;

exports[`removes vendor prefixes inside mediaQueries: react-test-renderer 1`] = `
@media (min-width: 200px) {
  .root {
    transform: translate(0, 0) !important;
  }
}

@supports (display: grid) {
  .root {
    display: flex !important;
  }
}

<div
  className="root"
/>
`;

exports[`removes vendor prefixes inside mediaQueries: react-testing-library 1`] = `
@media (min-width: 200px) {
  .root {
    transform: translate(0, 0) !important;
  }
}

@supports (display: grid) {
  .root {
    display: flex !important;
  }
}

<div
  class="root"
/>
`;
